lib/ansiblelint/rules/RunOnceRule.py

```diff
@@ -0,0 +1,67 @@
+
+# Copyright (c) 2020 Sorin Sbarnea <sorin.sbarnea@gmail.com>
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included in
+# all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+import sys
+
+from ansiblelint.rules import AnsibleLintRule
+
+
+class RunOnceRule(AnsibleLintRule):
+    """RunOnceRule Class."""
+
+    id = "107"
+    shortdesc = "Unpredictable run_once"
+    description = (
+        "Use of run_once does not work with free execution strategy. "
+        "Avoid it, or use noqa comments to ignore it.")
+    severity = 'MEDIUM'
+    tags = ['deprecated']
+    version_added = 'v4.4.0'
+
+    def matchtask(self, file, task):
+        return 'run_once' in task
+
+
+EXAMPLE_PLAYBOOK = """
+- hosts: localhost
+  strategy: free
+  tasks:
+
+    - name: task that mathes
+      command: touch /tmp/foo
+      run_once: true
+
+    - name: task that does not match
+      command: touch /tmp/foo
+      run_once: true  # noqa 107
+"""
+
+
+if "pytest" in sys.modules:
```

Comment:
This is... different. Do we want tests to live directly in the rule files, now?

Reply:
There's actually two other PRs attempting to make it possible/easy to couple tests with the rule modules. I still feel weird about it too...

Reply:
That is the goal, in fact the older PR also includes some documentation about how to write rules. Still we can wait for that to settle.

The main advantage is that I can use the example used in tests to create a rich documentation page for the rule.

My proposal is to experiment this approach with a limited set of rules and see how it scales. If we find deal-breakers, I will take care of moving the tests back.

Reply:
I like the idea. It can make it clear to me as a user what exactly the rule is testing for. I just hadn't looked into those other PRs.

I'm wondering if there is a more robust way for us to detect that we are running under our own tests? I'm just imagining a wild scenario where someone invokes ansible-lint while pytest is active and imported and somehow ends up running the test suite while trying to lint their own code. Is that safe from happening, or is there a way for us to make it safe? A custom environment variable, or command line argument, or something like that?

I just see this and immediately think it's far too easy for someone to have pytest autoloaded somehow and whoops, now they've got tests they didn't intend running.

Reply:
I think we do not need to worry about this because we do not support use of the linter other than from the command line. The presence of pylint in sys.modules has serious implications as loading pytest module does a lot of stuff under the hood.

Also keep in mind that under that conditional branch we only define test classes or method, not really executing stuff. So, even if you have pytest imported and you import the rule, it will not execute the tests.